### PR TITLE
enable defining global credentials in configurable namespace(s)

### PR DESCRIPTION
**Pull request by lrotim**
When setting up credentials for multiple kargo projects it would be practical to be able to put it in kargos namespace and have it available for multiple kargo projects.

Much like is the case with how argoCD enables storing creds in "his" namespace

### CODE REVIEW

File: charts/kargo/Chart.yaml
Comment (krancour) [inline charts/kargo/Chart.yaml:6]: These should never be touched... the release process sets them correctly.

**Comment by krancour** [comment]
@lrotim thanks for this!

I think a few of us maintainers will have to have a conversation about the security implications of this before making any decision.

**Comment by krancour** [comment]
Took some time to consider this over the weekend... adding my own thoughts to the discussion here:

1. I understand the need, but...

1. This would weaken the tenancy model, so...

1. It should be opt-in at install-time, just like Argo CD credential borrowing, but...

1. I don't like the idea of it being in Kargo's own namespace. I think most operating systems and Kubernetes all provide precedent for a separation between system-space and user-space. From my point of view, Kargo's own namespace is system-space and shouldn't be touched by users. Argo CD does _not_ play by these rules, but imho, that was a mistake that we shouldn't blindly emulate, so...

1. I would propose that in addition to being opt-in, that operators should designate a specific namespace to be used expressly for this purpose (shared credentials) at install-time.

@lrotim would this meet your needs?

@jessesuen wdyt?

**Comment by jessesuen** [comment]
The global credentials use case is valid, and I agree with the need to allow credentials that can be reused across projects. However, as Ken alluded to, this has a lot of security implications so I think the PR won't be able to go in as is.

> It should be opt-in at install-time, just like Argo CD credential borrowing, but...

The crux of the issue is: global credentials needs more controls. It won't be a simple true/false configuration but rather something like:

- https://github.com/my-org/* can be accessed by `*` kargo projects (this is probably the common case)

But it could also be:


> I don't like the idea of it being in Kargo's own namespace. I think most operating systems and Kubernetes all provide precedent for a separation between system-space and user-space. 

I don't see any issue with keeping the credentials in the kargo namespace, so long as the only user who can service/access these credentials kargo admin. Since this PR does not affect self-service

That said, we don't have to decide for the user. The implementation can/should be such that we don't explicitly require that they *must* reside in `kargo` namespace, but that we can reference the credentials explicitly. I suspect most people will choose to k


> Argo CD does not play by these rules, but imho, that was a mistake that we shouldn't blindly emulate, so...

Argo CD has a softer tenancy that required us to re-implement tenancy controls in API/controller. We've mostly avoided that by starting with Namespaces as projects, and how credentials are stored per project.

**Comment by jessesuen** [comment]
I filed https://github.com/akuity/kargo/issues/1048 for credential sharing / global credentials. As it stands, the approach of looking to `kargo` and then project namespace is too simplistic for what we want, and I think it will be better to spend more time to discuss a more proper/flexibile design.

**Comment by lrotim** [comment]
@krancour @jessesuen thx for the feedback. 

I understand this should take some consideration. I drafted a PR to be more concrete on the first steps.

I also considered pushing an update to move this from being explicitly kargo namespace, but figured I'd like to hear your feedback first and keep the changes to the minimum. 

I updated the PR to cover such behaviour now (not using kargo ns explicitly but allowing to define any number of namespaces for such purpose)

If you think that would be good enough "for now" with a TODO for wider considerations that would be great, but I understand if you  want to consider wider usecases and "wait" for a cleaner solution. 

Would be nice to have this as I'm trying to get my company to use kargo as early adopters. I've been tackling similar issues in the company and would prefer to use this as opposed to building "in-house". 

I've tried setting it up now, and current creds restrictions are a bit of a nuisance for the configuration I've setup, but we'll make due. I'll follow what you decide on this one

**Comment by lrotim** [comment]
Also, I agree - there should be some "access control" to specify which projects might access them. I wanted to keep the changes to the minimum and see if you would consider an option to add this for "admins" to share the creds, and design it in more detail later down the roadmap

**Comment by jessesuen** [comment]
@lrotim totally understand this current limitation is a big inconvenience. In https://github.com/akuity/kargo/issues/1048#issuecomment-1786080485 I likened the problem to the propagation of imagePullSecrets to different namespaces. I also mentioned that a workaround might be to use the same technique you are probably already using for imagePullSecrets, but for kargo credentials. Is that something that might work for you in the interim?

**Comment by lrotim** [comment]
@jessesuen Yes, for short/midterm we already have a workaround that will do, but wanted to try and contribute here. 
I'm happy to contribute/update this PR meet the spec defined in your linked issue, please let me know if this is something you would prefer to keep on official maintainers - if that is the case, we can close this up and just leave the link to the issue

File: internal/credentials/credentials.go
Comment (krancour) [inline internal/credentials/credentials.go:71]: Since there are multiple bits of configuration to include now, I would have a preference for starting to pass this config as an instance of a new type -- let's say `KubernetesDatabaseConfig`.

Something like this:

```golang
type KubernetesDBConfig struct {
	ArgoCDCredentialBorrowingEnabled bool     `envconfig:"ARGOCD_ENABLE_CREDENTIAL_BORROWING" default:"false"`
	ArgoCDNamespace                  string   `envconfig:"ARGOCD_NAMESPACE" default:"argocd"`
	SharedCredentialsNamespaces      []string `envconfig:"SHARED_CREDENTIALS_NAMESPACES" default:""`
}

func KubernetesDBConfigFromEnv() KubernetesDBConfig {
	cfg := KubernetesDBConfig{}
	envconfig.MustProcess("", &cfg)
	return cfg
}
```

Comment (krancour) [reply internal/credentials/credentials.go:71]: Also note that envconfig can parse comma-separated strings into a `[]string` for you. :smile:

File: internal/credentials/credentials.go
Comment (krancour) [inline internal/credentials/credentials.go:157]: Shouldn't we break out of the loop after finding credentials?

File: internal/credentials/credentials.go
Comment (krancour) [inline internal/credentials/credentials.go:128]: I think this whole block should be outdented one level from where it is. i.e. Close the previous `if secret == nil` block before this block.

Comment (krancour) [reply internal/credentials/credentials.go:128]: Mainy for readability.

File: charts/kargo/templates/controller/configmap.yaml
Comment (krancour) [inline charts/kargo/templates/controller/configmap.yaml:19]: For better or worse, we don't prefix env vars with `KARGO_`, so let's change this to `SHARED_CREDENTIALS_NAMESPACES`, please.

File: cmd/controlplane/controller.go
Comment (krancour) [inline cmd/controlplane/controller.go:147]: See [this comment](https://github.com/akuity/kargo/pull/1041/files#r1394755140)

This would become something like:

```golang
credentialsDB := credentials.NewKubernetesDatabase(
	credentials.KubernetesDatabaseConfigFromEnv(),
	kargoMgr.GetClient(),
	argoClientForCreds,
)
```

**Comment by krancour** [comment]
Ok... I agree with where this conversation went...

Let the "Kargo admin" pick namespace(s) for this purpose. If they pick `kargo`, fine -- that's their prerogative.

I think this PR moves the needle most of the way toward what #1048 is after and I think we should move ahead with it.

@lrotim you did a great job making this configurable.

I have a few nitpicky comments about implementation, but nothing major. Once they're set, let's merge this.

**Comment by jessesuen** [comment]
OK I took a close look at this, and I agree that this PR gets us as close as we need for the foreseeable future.

Introducing a `SHARED_CREDENTIALS_NAMESPACES` env var that a kargo admin can set at the system level, will solve the common desire to have global credentials defined *once*. And I don't think we need to solve the shared credential problem between only two projects (and not globally).

I realize now that the `SHARED_CREDENTIALS_NAMESPACES`  env var is effectively the same as my proposal to have search paths as described in https://github.com/akuity/kargo/issues/1048:

> A kargo admin can choose to configure "search paths" of namespaces for which to find credentials (similar to shell PATH env var) at the global level. When set at a global level, the search paths will apply to all projects. Global search paths are naturally opt-in by virtue of setting the system-wide search path to something.

So I'm good with the approach. I think we just need to fix some of the PR comments and I think this can be gtg!

File: charts/kargo/templates/controller/configmap.yaml
Comment (jessesuen) [reply charts/kargo/templates/controller/configmap.yaml:19]: I wonder if we should differentiate between what I'm calling "global" credentials and a different/future feature to share credentials between projects, not at the global scope. 

So I'm proposing we name this `GLOBAL_CREDENTIALS_NAMESPACES` since that's what this PR will enable.

**Comment by krancour** [comment]
@lrotim if you're time-constrained, we're happy to take this PR the rest of the way and give you the credit. You did most of the heavy lifting, after all. Just let us know.

**Comment by lrotim** [comment]
@krancour Thanks for the detailed feedback. I've had a working weekend, so I didn't have time to get back to this.

I wanted to touch upon this on the weekend and update the PR, taking comments into consideration due Monday. 

If you want to move this faster than that, I don't mind if you wanna wrap it up, will check on it on Saturday.

**Comment by krancour** [comment]
@lrotim I was only touching base as a courtesy in case you had a ton of other stuff on your plate. There's no rush.

**Comment by lrotim** [comment]
@krancour @jessesuen when you have the time, please check out if you find that changes are inlined with the comments

**Comment by krancour** [comment]
@lrotim I'm out sick today, but will have my 👀 on this first thing tomorrow. I can't say enough how much we appreciate this.

File: charts/kargo/values.yaml
Comment (krancour) [inline charts/kargo/values.yaml:185]: @lrotim, I think it would feel more natural for this to be a list of strings instead of a single, comma-delimited string.

File: cmd/controlplane/controller.go
Comment (krancour) [inline cmd/controlplane/controller.go:139]: I don't think the client should be treated as an option, because it's 100% required. Could we have two args: client and opts?

Comment (krancour) [reply cmd/controlplane/controller.go:139]: Sorry... I mis-spoke here.

The "Kargo client" is 100% required. This one is optional, depending on whether credential borrowing is enabled.

So... imho, treat the "Kargo client" as required and leave the "Argo client" in the realm of opts.

Comment (lrotim) [reply cmd/controlplane/controller.go:139]: In practice, I've seen a lot of "opts" usage for both required and "optional args".  

But makes sense if you wanna make this explicit

**Review by krancour — APPROVED** [review]

**Pull request merged**
